src/fastpath/fastpath.c

```diff
@@ -902,3 +902,240 @@ void NORETURN fastpath_vm_fault(vm_fault_type_t type)
     fastpath_restore(badge, msgInfo, NODE_STATE(ksCurThread));
 }
 #endif
+
+#ifdef CONFIG_IRQ_FASTPATH
+#ifdef CONFIG_ARCH_ARM
+static inline
+FORCE_INLINE
+#endif
+void NORETURN fastpath_irq()
+{
+
+    sched_context_t *sc = NULL;
+    bool_t schedulable = false;
+    bool_t idle = false;
+    tcb_t *dest = NULL;
+    cap_t newVTable;
+    vspace_root_t *cap_pd;
+    pde_t stored_hw_asid;
+
+    /* The IRQ should be passed as a parameter to this fn. At the moment, if the
+    slowpath is called, calls getActiveIRQ again in handleInterruptEntry.
+    Syscall API change to handleInterruptEntry will be done in another commit. */
+    irq_t irq = getActiveIRQ();
+
+    /* check the irq is valid */
+    if (unlikely(IRQT_TO_IRQ(irq) == IRQT_TO_IRQ(irqInvalid))) {
+#ifdef CONFIG_IRQ_REPORTING
+        printf("Spurious interrupt\n");
+#endif
+        handleSpuriousIRQ();
+        restore_user_context();
+        UNREACHABLE();
+    }
+
+    if (unlikely(IRQT_TO_IRQ(irq) > maxIRQ)) {
+        maskInterrupt(true, irq);
+        ackInterrupt(irq);
+        restore_user_context();
+        UNREACHABLE();
+    }
+
+    /* only fastpath signalling IRQs*/
+    if (unlikely(intStateIRQTable[IRQT_TO_IDX(irq)] != IRQSignal)) {
+        slowpath_irq();
+        UNREACHABLE();
+    }
+
+    cap_t ntfn_cap = intStateIRQNode[IRQT_TO_IDX(irq)].cap;
+    if (unlikely(cap_get_capType(ntfn_cap) != cap_notification_cap ||
+                !cap_notification_cap_get_capNtfnCanSend(ntfn_cap))) {
+#ifdef CONFIG_IRQ_REPORTING
+        printf("Undelivered irq: %d\n", (int) IRQT_TO_IRQ(irq));
+#endif
+        maskInterrupt(true, irq);
+        ackInterrupt(irq);
+        restore_user_context();
+        UNREACHABLE();
+    }
+
+    notification_t *ntfn_ptr = NTFN_PTR(cap_notification_cap_get_capNtfnPtr(ntfn_cap));
+    notification_state_t ntfn_state = notification_ptr_get_state(ntfn_ptr);
+    word_t badge = cap_notification_cap_get_capNtfnBadge(ntfn_cap);
+
+    switch (ntfn_state) {
+    case NtfnState_Active:
+#ifdef CONFIG_BENCHMARK_TRACK_KERNEL_ENTRIES
+        ksKernelEntry.is_fastpath = true;
+#endif
+        ntfn_set_active(ntfn_ptr, badge | notification_ptr_get_ntfnMsgIdentifier(ntfn_ptr));
```

Comment:
Slowpath only does `notification_ptr_set_ntfnMsgIdentifier` if ntfn is already active

Reply:
Yeah, i think you're right. I think it is benign to do it this way since setting it to active again should essentially be a no-op, but it is unnecessary. Just `notification_ptr_set_ntfnMsgIdentifier` should suffice.